### PR TITLE
Add pagination support to `usePromise`, `useCachedPromise` and `useFetch`.

### PR DESCRIPTION
**Pull request by sxn**
### Context
We're working on adding pagination support to `Grid` and `List` via a new `pagination` prop. To support users in more easily taking advantage of pagination, this PR updates `usePromise`, `useCachedPromise` and `useFetch` to have pagination support. This means a couple of things:
1. The hooks' return signature is expanded to also include `pagination: { hasMore: boolean, pageSize: number, onLoadMore: () => void }`, which can be directly passed to the `List` or `Grid` component
2. The hooks' signature is expanded to accept "paginated" variants of their respective arguments. 

#### For `usePromise` and `useCachedPromise`

the hooks' first argument can be not just a function that returns a promise, but a function that returns a paginated promise. A paginated promise in this case is a function that receives pagination info and returns a promise. It means going from 

```TSX
const { data, isLoading } = usePromise(
  async (searchText: string) => { 
    const data = … // data can be anything
    return data;
  },
  [searchText]
)
``` 

to

```TSX
const { data, isLoading, pagination } = usePromise(
  (searchText: string) => async (options: { page: number; lastItem?: string }) => { 
    const data = … // data _needs_ to be an array
    const hasMore = …
    return { data, hasMore };
  },
  [searchText]
);
```

Particular to `useCachedPromise`: when paginating, only the first page for each pagination instance is cached.

#### For `useFetch`

the hook's first argument can be not just a `RequestInfo`, but a paginated request info. 
Similar to above, this means going from:

```TSX
const { data, isLoading } = useFetch(url);
```

to

```TSX
type FetchResponse = { … } // whatever the API call's expected return type is
type Item = { … }
const { data, isLoading, pagination } = useFetch(
  (options: { page: number, lastItem?: any }) => url,
  {
    mapResult: (result: FetchResponse) {
      const data = … // data _needs_ to be an array
      const hasMore = …
      return { data, hasMore };
    }
  }
);
```

The main difference between `usePromise` / `useCachedPromise` and `useFetch` is that the latter _requires_ a `mapResult` option to be passed, in addition to the paginated request info, in order to tell the hook whether it should try to paginate further, and to have a chance to "process" the data returned by the API, in case e.g. the API returns something like `{ success: boolean, results: Item[] }`.

#### Other changes

In order to support these changes, I've had to make a couple more changes:
- `onData` now receives not just the `data`, but also a second argument: the pagination options. 
  - the reason for this change is: pagination "data", i.e. `page`, latest `hasMore`, `lastItem` and the data accumulated from the current pagination instance, it's all stored in `usePromise`. However, because `useCachedPromise` only caches the first page, we needed a way to figure out what page we're currently at.
- `useCachedPromise` and `useFetch` accept a `cacheKeySuffix` option, which, as the name implies will be added to the cacheKey that `useCachedPromise` already generates internally. Because `useFetch` builds on top of `useCachedPromise`, and because `useCachedPromise` uses the function & the args to generate the cache key – this wasn't enough to differentiate, when using e.g. `useFetch((searchText: string) => async (options) => "https://example.com/?page= " + options.page + "&searchText=" + searchText, [searchText])`, between `searchText` being empty and not, i.e.: 

https://github.com/raycast/utils/assets/1155589/0eefe9b3-b922-495d-9c6b-08acd05ca9aa

### Screenshots

**usePromise**:

https://github.com/raycast/utils/assets/1155589/735a5ab6-a51e-4af6-b149-89b62c47996d

**useCachedPromise**:

https://github.com/raycast/utils/assets/1155589/050c1e54-c091-49a6-8e49-b31f8be6f975

**useFetch**:

https://github.com/raycast/utils/assets/1155589/b3984890-4c14-4e79-9952-e54c685cc328

### CODE REVIEW

File: src/useCachedPromise.ts [hunk 1]
```diff
@@ -20,6 +30,11 @@ export type CachedPromiseOptions<T extends FunctionReturningPromise, U> = Promis
    * This is particularly useful when used for data for a List to avoid flickering.
    */
   keepPreviousData?: boolean;
+  /**
+   * The hook generates a cache key from the promise & its arguments. Sometimes that's not enough to guarantee,
+   * uniqueness, and in those cases you can pass a `cacheKeySuffix`.
+   */
+  cacheKeySuffix?: string;
```
Comment (mathieudutour) [inline src/useCachedPromise.ts:37]: I'm wondering if we have to expose it 🤔

File: src/useFetch.ts [hunk 2]
```diff
@@ -68,15 +72,41 @@ async function defaultParsing(response: Response) {
...
+type PaginatedRequestInfo = (pagination: { page: number; lastItem?: any }) => RequestInfo;
+
+export function useFetch<V = unknown, U = undefined, T = V>(
   url: RequestInfo,
-  options?: RequestInit & { parseResponse?: (response: Response) => Promise<T> } & Omit<
-      CachedPromiseOptions<(url: RequestInfo, options?: RequestInit) => Promise<T>, U>,
-      "abortable"
-    >,
+  options?: RequestInit & {
+    parseResponse?: (response: Response) => Promise<T>;
```
Comment (mathieudutour) [inline src/useFetch.ts:80]: how about accepting `mapResult` here as well? It's useful even without pagination

File: tests/src/fetch-paginated.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        };
+      },
+      keepPreviousData: true,
+      initialData: [],
+    },
+  );
+
+  return (
+    <List isLoading={isLoading} pagination={pagination} onSearchTextChange={setSearchText}>
+      {data?.map((company) => (
```
Comment (mathieudutour) [inline tests/src/fetch-paginated.tsx:59]:
```suggestion
      {data.map((company) => (
```

File: tests/src/cached-promise-paginated.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      isLoading={isLoading}
+      onSearchTextChange={setSearchText}
+      pagination={pagination}
+      actions={
+        <ActionPanel>
+          <Action title="Reload" onAction={() => revalidate()} />
+        </ActionPanel>
+      }
+    >
+      {data?.map((item) => <List.Item key={item} title={item} />)}
```
Comment (mathieudutour) [inline tests/src/cached-promise-paginated.tsx:33]:
```suggestion
      {data.map((item) => <List.Item key={item} title={item} />)}
```

File: src/useCachedPromise.ts (hunk 1 above)
Comment (sxn) [reply src/useCachedPromise.ts:37]: we don't _need_ to expose it, no. I'll make it internal-only for now.

File: src/types.ts [hunk 5]
```diff
@@ -1,4 +1,19 @@
-export type FunctionReturningPromise<T extends any[] = any[]> = (...args: T) => Promise<any>;
+export type PaginationOptions<T = any> = {
+  page: number;
```
Comment (mathieudutour) [inline src/types.ts:2]: worth adding some JS docs here? Maybe saying that it's a 0-based index

**Review by mathieudutour — APPROVED** [review]
Looks good, the only thing missing is the peerDependency update in the package.json (which I assume you'll have to wait until the release?)

**Comment by mathieudutour** [comment]
ah you're missing the doc updates as well

**Pull request merged**